### PR TITLE
Add arcface head

### PR DESCRIPTION
**Pull request by zzc98**
## Motivation

Support [ArcFace loss](https://arxiv.org/abs/1801.07698).

## Checklist

**Before PR**:

- [ ] Pre-commit or other linting tools are used to fix the potential lint issues.
- [ ] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects, like MMDet or MMSeg.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

**Comment by okotaku** [comment]
@zzc98 @Ezra-Yu 

Thank you for creating useful components!
I have other arcface implementation and I prefer it.
Please check it too.

https://github.com/open-mmlab/mmclassification/pull/933

File: mmcls/models/heads/arcface_head.py [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+                data of every samples. If not None, set ``pred_label`` of
+                the input data samples. Defaults to None.
+
+        Returns:
+            torch.Tensor: features of samples.
+        """
+
+        pre_logits = self.pre_logits(feats)
+        if self.used == 'after':
+            pre_logits = self.norm_layer(pre_logits)
```
Comment (okotaku) [inline mmcls/models/heads/arcface_head.py:214]: output from norm layer is scaled. 

```
pre_logits = self.s * self.norm_layer(pre_logits)
```

File: mmcls/models/heads/arcface_head.py [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+
+        pre_logits = self.pre_logits(feats)
+
+        # cos=(a*b)/(||a||*||b||)
+        cosine = self.norm_layer(pre_logits)
+
+        if target is None:
+            return cosine
+
+        # sin^2+cos^2=1
```
Comment (okotaku) [inline mmcls/models/heads/arcface_head.py:133]: I think former implementation is simpler than latter.

```
phi = torch.cos(torch.acos(cosine) + self.m)
```

```
sine = torch.sqrt(1.0 - torch.pow(cosine, 2))
phi = cosine * self.cos_m - sine * self.sin_m
```

**Review by okotaku — CHANGES_REQUESTED** [review]

File: mmcls/models/heads/arcface_head.py (hunk 1 above)
Comment (zzc98) [reply mmcls/models/heads/arcface_head.py]: Thanks for your reminding, the modification has been made.

**Review by okotaku — APPROVED** [review]
LGTM

File: mmcls/models/heads/arcface_head.py (hunk 2 above)
Comment (wang11wang) [reply mmcls/models/heads/arcface_head.py]: Though the former is simpler than the latter,
But the former implementation may suffers numerical problems when the value is very small with auto mixed-precision training

Comment (okotaku) [reply mmcls/models/heads/arcface_head.py]: @wang11wang
Can you give an example of a calculation, which may cause problems?

**Review by Ezra-Yu — APPROVED** [review]
LGTM.
